kvprober: rate limit the planner

joshimhoff opened the pull request:
https://github.com/cockroachdb/cockroach/issues/61255

**kvprober: rate limit the planner**

This commit introduces a planning rate limit. This protects CRDB from
planning executing too often, due to either issues with CRDB (e.g.
meta2 unavailability) or bugs in kvprober. When planning does execute,
kvprober scans kv.prober.planner.num_steps_to_plan_at_once rows worth
of meta2 and unmarshalls the resulting range descriptors.

Release justification: Auxiliary system that is off by default.
Release note: None.

joshimhoff committed:
kvprober: rate limit the planner

This commit introduces a planning rate limit. This protects CRDB from
planning executing too often, due to either issues with CRDB (e.g.
meta2 unavailability) or bugs in kvprober. When planning does execute,
kvprober scans kv.prober.planner.num_steps_to_plan_at_once rows worth
of meta2 and unmarshalls the resulting range descriptors.

Release justification: Auxiliary system that is off by default.
Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61275)
<!-- Reviewable:end -->

tbg reviewed (commented):
Random note: I missed earlier that we are using `time.Ticker` here, I think you want to use `timeutil.NewTimer` and after reading from `t.C` you want to do `t.Read = true`. Shouldn't change anything, but it's cleaner.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-:-MUnCmHBBhQnhLL-mYU5:bq3rxlq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @joshimhoff, @jreut, and @knz)

___
*[pkg/kv/kvprober/planner.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-MUnClTfAy4W9ill7WZm:-MUnClTfAy4W9ill7WZn:bcf7g3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a463bd04c544864108036d5082a3d710cf3eb6/pkg/kv/kvprober/planner.go#L91)):*
> ```
> .probe
> ```
___
*[pkg/kv/kvprober/settings.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-MUnBGbe5MuytDevdyJK:-MUnBGbf8gmKDP80xB13:bxmmppr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6430b4d9c0c1f2ad7aba8e6bf476ffa986433b1/pkg/kv/kvprober/settings.go#L78)):*
> ```Go
> 	})
> 
> var plannerRateLimit = settings.RegisterDurationSetting(
> ```

Hmm. I know I was on board with this approach, but now that I see it it seems less simple than I anticipated. For one, with the cluster setting, we now have to wonder what the behavior is when you set it to a value smaller than the probing interval (I think it will just bounce a few probes every time it needs to do planning and pollute the metrics, so not too bad?)

I think the right thing here could be to avoid the cluster setting and to "hard"-code it as 0.5 times the probing interval times the prefetch size.
So if you're fetching up to 100 descs during planning, and you want to run a step every 2s, if you ever hit an error while fetching descs you would back off 0.5 * 100 * 2s = 100s. Or, in other words, the load on meta2 that results from a failure is at most 2x the load it would get if probing were to occur normally. As a result the meta ranges are protected and we don't have to introduce another knob that could be tuned incorrectly. We could even just drop the random factor `0.5` that I threw in there.


<!-- Sent from Reviewable.io -->

joshimhoff reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @joshimhoff, @jreut, @knz, and @tbg)

___
*[pkg/kv/kvprober/settings.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-MUnBGbe5MuytDevdyJK:-MUnHdyjAVvjLozQey9Y:b-o52xkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6430b4d9c0c1f2ad7aba8e6bf476ffa986433b1/pkg/kv/kvprober/settings.go#L78)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm. I know I was on board with this approach, but now that I see it it seems less simple than I anticipated. For one, with the cluster setting, we now have to wonder what the behavior is when you set it to a value smaller than the probing interval (I think it will just bounce a few probes every time it needs to do planning and pollute the metrics, so not too bad?)

I think the right thing here could be to avoid the cluster setting and to "hard"-code it as 0.5 times the probing interval times the prefetch size.
So if you're fetching up to 100 descs during planning, and you want to run a step every 2s, if you ever hit an error while fetching descs you would back off 0.5 * 100 * 2s = 100s. Or, in other words, the load on meta2 that results from a failure is at most 2x the load it would get if probing were to occur normally. As a result the meta ranges are protected and we don't have to introduce another knob that could be tuned incorrectly. We could even just drop the random factor `0.5` that I threw in there.
</blockquote></details>

I think I like that better too. Thanks for the idea. Gonna try it out.


<!-- Sent from Reviewable.io -->

joshimhoff reviewed (commented):
PTAL!

I'm gonna add moving to `timeutil.NewTicket` to my list of maintenance work to get done post 21.1 branch being cut. It's not totally clear to me what you mean re: `t.Read` and I want to get one other change into 21.1 ideally (add structured logging) and also do some e2e manual testing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-:-MUnJCnC6UJDkUJA4CiR:b-p0w20m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jreut, @knz, and @tbg)

___
*[pkg/kv/kvprober/planner.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-MUnClTfAy4W9ill7WZm:-MUnJ8VM12pAbOjlA4dH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a463bd04c544864108036d5082a3d710cf3eb6/pkg/kv/kvprober/planner.go#L91)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> ```
> .probe
> ```

</blockquote></details>

Done.
___
*[pkg/kv/kvprober/settings.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-MUnBGbe5MuytDevdyJK:-MUnJBLQBUh-21CEwl25:b-13hihi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6430b4d9c0c1f2ad7aba8e6bf476ffa986433b1/pkg/kv/kvprober/settings.go#L78)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

I think I like that better too. Thanks for the idea. Gonna try it out.
</blockquote></details>

Done. I like it.


<!-- Sent from Reviewable.io -->

tbg committed:
kvprober: use Timer instead of Ticker

Release justification: bug fixes and low-risk updates to new functionality
Release note: None

tbg committed:
kvprober: use a test planning rate limit of zero

A nonzero limit has the potential to cause issues, albeit theoretical
ones, if the clock doesn't advance even 1ns between two operations.
This is mostly hypothetical and with zero we still have issues if the
clock jumps backwards, but still.

Release justification: non-production code changes
Release note: None

tbg reviewed (approved):
Looks good! I added two commits to your branch. You should be able to merge at will.

bors d+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61275#-:-MUsqpmj-XVO67F92O4M:bjjjn0a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jreut and @knz)



<!-- Sent from Reviewable.io -->

joshimhoff commented:
TTTF and two commits, Tobias!

joshimhoff commented:
bors r+
